plugin/src/main/java/com/denizenscript/denizen/objects/properties/entity/EntityAttributeModifiers.java

```diff
@@ -81,27 +84,64 @@ public static MapTag mapify(AttributeModifier modifier) {
+        }
+        if (NMSHandler.getVersion().isAtLeast(NMSVersion.v1_21)) {
+            result.putObject("key", new ElementTag(Utilities.namespacedKeyToString(modifier.getKey()), true));
+        }
+        // TODO: remove/deprecate these two
         result.putObject("slot", new ElementTag(modifier.getSlot() == null ? "any" : modifier.getSlot().name()));
-        if (NMSHandler.getVersion().isAtMost(NMSVersion.v1_20)) {
+        try {
             result.putObject("id", new ElementTag(modifier.getUniqueId().toString()));
         }
+        catch (Exception ignored) {}
         return result;
     }
 
-    public static AttributeModifier modiferForMap(Attribute attr, MapTag map) {
-        ElementTag name = map.getElement("name");
+    public static AttributeModifier modiferForMap(Attribute attr, MapTag map, TagContext context) {
         ElementTag amount = map.getElement("amount");
         ElementTag operation = map.getElement("operation");
-        ElementTag slot = map.getElement("slot", "any");
-        // TODO: MC 1.21: ID and name are now the same internal value.
-        ElementTag id = map.getElement("id");
-        UUID idValue;
         double amountValue;
         AttributeModifier.Operation operationValue = operation.asEnum(AttributeModifier.Operation.class);
         if (operationValue == null) {
             Debug.echoError("Attribute modifier operation '" + operation + "' does not exist.");
             return null;
         }
+        try {
+            amountValue = Double.parseDouble(amount.toString());
+        }
+        catch (NumberFormatException ex) {
+            Debug.echoError("Attribute modifier amount '" + amount + "' is not a valid decimal number.");
+            return null;
+        }
+        ElementTag key = map.getElement("key");
+        boolean is1_21 = NMSHandler.getVersion().isAtLeast(NMSVersion.v1_21);
+        if (!is1_21) {
+            return parseLegacyModifier(attr, map, amountValue, operationValue);
+        }
+        if (key == null && map.size() > 2) {
+            BukkitImplDeprecations.pre1_21AttributeFormat.warn(context);
+            return parseLegacyModifier(attr, map, amountValue, operationValue);
+        }
+        if (key == null) {
+            Debug.echoError("Must specify a key.");
+            return null;
+        }
+        String groupName = map.getElement("slot_group", "any").asString();
+        EquipmentSlotGroup group = EquipmentSlotGroup.getByName(groupName);
+        if (group == null) {
+            Debug.echoError("Invalid equipment slot group specified: " + groupName);
+            return null;
+        }
+        return new AttributeModifier(Utilities.parseNamespacedKey(key.asString()), amountValue, operationValue, group);
+    }
+
+    private static AttributeModifier parseLegacyModifier(Attribute attr, MapTag map, double amount, AttributeModifier.Operation operation) {

```

Comment:
should not be `private`

Reply:
Made it public, but deprecated it for removal (as it uses API that's deprecated for removal, and any code using it should ideally update)